Allow separators to be omitted from lists.

nate-chandler opened the pull request:
Enables newlines to be used in place of separators in some lists.

nate-chandler committed:
WIP: Allow separators to be omitted from lists.

nate-chandler commented:
@swift-ci please test source compatibility

harlanhaskins commented:
@swift-ci please test source compatibility

nate-chandler committed:
Basic test coverage.

AnnaZaks commented:
@swift-ci please test source compatibility

nate-chandler committed:
Allowed omission from function calls and tuples.

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
Added tests on initializers.

nate-chandler committed:
Corrected argument capitalization.

nate-chandler commented:
@swift-ci please test source compatibility

benlangmuir committed:
Add tests for indentation, swift-syntax for optional commas

benlangmuir committed:
Add code-completion tests for optional commas

benlangmuir commented:
@nate-chandler I wrote some tests for the IDE side; it looks like everything I've thought about is already working, which is great.  I pushed a `omit-comma-ide-tests` branch to my swift fork: https://github.com/benlangmuir/swift/tree/omit-comma-ide-tests . You could probably just cherry-pick the two test commits at the top of that branch into this PR.

nate-chandler commented:
@benlangmuir Fantastic! Will do.

nate-chandler commented:
@benlangmuir Okay, the commits are cherry-picked to 
https://github.com/apple/swift/pull/22714/commits/cf8c2f6ea9252ae139fb048942467e2e9ec6b874 and 
https://github.com/apple/swift/pull/22714/commits/97cfe0eb01f8118717a7794e41d92fabfd17e94a .

benlangmuir commented:
@swift-ci please test

nate-chandler committed:
Corrected expected error messages in Parse/invalid.swift.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Stmts/decls can start line if separators omitted.

Allowing them here enables function calls with argument labels whose
names are contextual keywords that introduce stmts (for) and decls
(protocol).

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Bail out of expr sequence parsing if peek shows :.

If we're parsing an expression sequence and we're about to hit a colon,
our expression is going to end before that colon so bail out before
consuming the preceeding token.

nate-chandler committed:
Updated tests with revised diagnostics.

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
Updated new parseExprList call.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

benlangmuir committed:
Fix test after change to contextual type

This case was previously getting type [Any], but is now [T] so update
the output to match. It would be better to recover as [S], but that is
an existing bug with type context for array literals.

benlangmuir commented:
@swift-ci please test source compatibility

benlangmuir commented:
@swift-ci please test

nate-chandler commented:
@swift-ci Please test Linux platform

nate-chandler commented:
@swift-ci Please clean test Linux platform

xedin reviewed (changes requested):
LGTM! Besides logic in `parseList` for which I've left comment inline.

nate-chandler commented on lib/Parse/Parser.cpp:
```diff
@@ -1016,7 +1017,7 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
     // If we're in a comma-separated list, the next token is at the
     // beginning of a new line and can never start an element, break.
-    if (Tok.isAtStartOfLine() &&
+    if (Tok.isAtStartOfLine() && !AllowSepOmission &&
```
I'm not sure.  My goal with this change was to enable commas to be omitted from argument lists whose argument labels are contextual keywords.  Is there a better way to achieve that?

xedin replied:
What do you mean by contextual keywords? I'm just trying to point out if there is an element on the new line but it's something which is not allowed to be in the list in general e.g. declaration or statement, loop should stop just like before. Comma omitting should only be allowed if next token represents something which could syntactically belong in the list...

nate-chandler replied:
`protocol`, for example, can be used as an argument label:

```swift
func foo(a: Int, protocol: Int, c: Int) {}
```

It would be nice to be able to omit the comma before the second item in the list `protocol: 16`:

```swift
foo(
    a: 13
    protocol: 16
    c: 15
)

```

Without `!AllowSepOmission &&` or a better change, that fails to compile because `isStartOfDecl()` returns true.

```
main.swift:28:5: error: expected ')' in expression list
    protocol: 16
    ^
main.swift:26:4: note: to match this opening '('
foo(
   ^
main.swift:27:10: error: missing argument for parameter 'protocol' in call
    a: 13
         ^
         , protocol: <#Int#>
main.swift:24:6: note: 'foo(a:protocol:)' declared here
func foo(a: Int, protocol: Int) {}
     ^
main.swift:28:13: error: expected identifier in protocol declaration
    protocol: 16
            ^
```

Another approach would be to modify `isStartOfDecl()` to look ahead farther.  Would that be a better solution?  It seems like it might be enough to peek one token after the keyword and if it's a colon, return `false`--looking at the list of keywords in `isKeywordPossibleDeclStart()`, it seems that none of them can be followed immediately by a colon.  Is that right?

That alternate change corrects the regression in `test/decl/subscript/subscripting.swift`.

nate-chandler replied:
I definitely see your point that if the item in the list is 

> something which is not allowed to be in the list in general e.g. declaration or statement, loop should stop just like before

so it seems that enhancing `isStartOfDecl()` and `isStartOfStmt()` to be slightly more discerning seems like a much better solution.  The loops stops like before when the next item is a declaration or statement.

nate-chandler replied:
Fixed in https://github.com/apple/swift/pull/22714/commits/acdd16f7d16d1d2223758b09f5327175f1c3f8d6 .

xedin replied:
I agree that the best solution, as you suggested, would be indeed to do  lookahead at what is going on to identify label syntax. It seems like `isStartOfDecl` should know that `identifier:` doesn't constitute a valid declaration (unless it starts with `case`). There is  `canBeArgumentLabel()` method which could be used in combination with `peekToken.is(tok::colon)` just like what `Parser::parseOptionalArgumentLabel` does to determine that.

nate-chandler replied:
Ok, as of https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be , `canBeArgumentLabel()` is used alongside `peekToken().is(tok::colon)`, and it's all done in `parseList`.

nate-chandler committed:
Parsed keywords as argument labels correctly, fixing diagnostic regression.

nate-chandler commented on test/decl/subscript/subscripting.swift:
```diff
@@ -304,11 +304,10 @@ func testSubscript1(_ s1 : SubscriptTest1) {
   // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}}
   _ = s1.subscript("hello"
   // expected-error@-1 {{value of type 'SubscriptTest1' has no property or method named 'subscript'; did you mean to use the subscript operator?}}
-  // expected-note@-2 {{to match this opening '('}}
```
Fixed in https://github.com/apple/swift/pull/22714/commits/acdd16f7d16d1d2223758b09f5327175f1c3f8d6 .

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci Please clean test Linux platform

xedin commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -2588,6 +2589,8 @@ bool Parser::isStartOfDecl() {
   // If this is obviously not the start of a decl, then we're done.
   if (!isKeywordPossibleDeclStart(Tok)) return false;
 
+  if (peekToken().is(tok::colon)) return false;
```
As I suggested in my other comment this should probably be - `if (Tok.canBeArgumentLabel() && peekToken().is(tok::colon)) return false;`.

xedin replied:
Safer change might be to add that logic to `parseList` when it's detected that token is at the newline and separator omission is allowed.

nate-chandler replied:
In https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be , the logic you described here is used and the change to `isStartOfStmt` and `isStartOfDecl` is reverted.

xedin commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3451,6 +3465,9 @@ ParserResult<Expr> Parser::parseExprCollection() {
         break;
 
+    if (Tok.isAtStartOfLine())
+        continue;
```
Small nit: we use 2 spaces for indentation.

nate-chandler replied:
Fixed in https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be .

xedin commented on lib/Parse/Parser.cpp:
```diff
@@ -1016,7 +1017,7 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
     // If we're in a comma-separated list, the next token is at the
     // beginning of a new line and can never start an element, break.
-    if (Tok.isAtStartOfLine() &&
+    if (Tok.isAtStartOfLine() && 
```
nit: whitespace.

nate-chandler replied:
Fixed in https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be .

xedin commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -62,6 +62,7 @@ bool Parser::isStartOfStmt() {
   case tok::pound_warning:
   case tok::pound_error:
   case tok::pound_sourceLocation:
+    if (peekToken().is(tok::colon)) return false;
```
`default:` is a valid statement.

nate-chandler replied:
Reverted in https://github.com/apple/swift/pull/22714/commits/faecb8d5f15a3421f661bd606158d0394dbb19be because the logic is now in `parseList` as you described.

nate-chandler committed:
Lookahead in parseList, not isStartOf{Decl,Stmt}.

Fixed whitespace.

nate-chandler commented:
@swift-ci please clean test

xedin commented on lib/Parse/Parser.cpp:
```diff
@@ -1015,9 +1015,19 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
+    // argument label and continue parsing the list.
+    if (Tok.isAtStartOfLine() && AllowSepOmission && Tok.canBeArgumentLabel() 
+        && peekToken().is(tok::colon)) { 
+      continue;
```
Optional: It would be great to unify that into a single block `if (Tok.isAtStartOfLine()) { if (allowed && cond) continue; if (invalid) break; }` if possible.

nate-chandler replied:
Changed to

```
    if (Tok.isAtStartOfLine()) {
      if (AllowSepOmission && Tok.canBeArgumentLabel()
          && peekToken().is(tok::colon))
        // If the next token is at the beginning of a new line, could be an
        // argument label, is followed by a colon, and separator omission is
        // permitted, expect that we'll parse another list element which begins
        // with that argument label and continue parsing the list.
        continue;
      else if (Tok.is(tok::r_brace) || isStartOfDecl() || isStartOfStmt())
        // If we're in a comma-separated list, the next token is at the
        // beginning of a new line and can never start an element, break.
        break;
    }
``` 

in https://github.com/apple/swift/pull/22714/commits/4b8c2f9b06e64ab2415febdb840f2781220ec707 . Is this what you have in mind?

xedin replied:
Yes, thank you!

xedin reviewed (approved):
LGTM!

nate-chandler committed:
Unified two separate conditional blocks into one.

nate-chandler commented:
@swift-ci please clean test

nate-chandler commented:
@swift-ci please test source compatibility

xedin commented on lib/Parse/Parser.cpp:
```diff
@@ -1016,20 +1016,18 @@ Parser::parseList(tok RightK, SourceLoc LeftLoc, SourceLoc &RightLoc,
-      break;
+    if (Tok.isAtStartOfLine()) { 
+      if (AllowSepOmission && Tok.canBeArgumentLabel() 
+          && peekToken().is(tok::colon)) 
```
One more nit I glanced over: we use `clang-format` which makes this code incorrectly formatted. The workflow I have to add run git-clang-format on the files I have ready to be committed to make sure everything is correctly formatted. The tool is shipped together with our swift-clang so you just need to add `<swift-dir>/clang/tools/clang-format` to your `PATH` to get that working.

nate-chandler replied:
Thanks!  I'll set my system up that way.  Fixed the whitespace in https://github.com/apple/swift/pull/22714/commits/2adc41d697485f395c14294a2f1b8e8b796e32e1 .

nate-chandler committed:
Fixed whitespace.

tkremenek commented:
@swift-ci build toolchain

tkremenek commented:
@swift-ci clean build toolchain

tkremenek commented:
@swift-ci clean build

nate-chandler commented:
Toolchains:

macOS: https://ci.swift.org/job/swift-PR-toolchain-osx/272/artifact/branch-master/swift-PR-22714-272-osx.tar.gz
https://ci.swift.org/job/swift-PR-toolchain-Linux/211/artifact/branch-master/swift-PR-22714-211-ubuntu16.04.tar.gz

hartbit commented:
I tried the toolchain and it doesn't seem to allow me eliding commas in parameter lists

nate-chandler commented:
> I tried the toolchain and it doesn't seem to allow me eliding commas in parameter lists

Hi @hartbit,

Thanks for the report!  What you're seeing is intentional behavior.  The proposed change allows comma elision only in expression lists.

```
func foo(
    _ a: Int, // required here
    _ b: Int
)
    -> Int
{
    return a + b
}

let s = foo(
    1 // optional here
    2
)
```

brentdax commented:
@swift-ci please build toolcahin

brentdax commented:
Trying to build a toolchain with this feature and static subscripts.
